File: MVP/app/src/main/java/com/example/mvp/main/db/ProductLocalDataSourceImp.java
package com.example.mvp.main.db;

import android.content.Context;

import androidx.lifecycle.LiveData;

import com.example.mvp.main.model.Product;

import java.util.List;

public class ProductLocalDataSourceImp implements ProductLocalDataSource {



    private ProductDAO productDAO ;
    private static ProductLocalDataSourceImp localSource = null ;
    private LiveData<List<Product>> products ;

    public ProductLocalDataSourceImp(Context context) {
        productDAO = AppDataBase.getInstance(context).getProductDAO();
        products = productDAO.getAllProducts();
    }

    public static ProductLocalDataSourceImp getInstance(Context context) {
        if (localSource == null)
        {
            localSource = new ProductLocalDataSourceImp(context);
        }

        return localSource ;
    }

    @Override
    public LiveData<List<Product>> getAllProducts() {
        return products;
    }

    @Override
    public void insertProduct(Product product) {
        new Thread()
        {
            @Override
            public void run() {
                productDAO.insertProduct(product);
            }
        }.start();

    }

    @Override
    public void deleteProduct(Product product) {

        new Thread()
        {
            @Override
            public void run() {
                productDAO.deleteProduct(product);
            }
        }.start();



    }
}

File: MVVM/app/src/main/java/com/example/mvvm/model/db/ProductLocalDataSourceImp.kt
package com.example.mvvm.model.db

import android.content.Context
import com.example.mvvm.model.Product
import com.example.mvvm.model.ProductLocalDataSource
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class ProductLocalDataSourceImp private constructor(context: Context) : ProductLocalDataSource {

    companion object {
        private lateinit var dao: ProductDao

        @Volatile
        private var instance: ProductLocalDataSourceImp? = null

        fun getInstance(context: Context): ProductLocalDataSourceImp {
            return instance ?: synchronized(this)
            { instance ?: ProductLocalDataSourceImp(context).also{
                dao = ProductDataBase.getInstance(context).getProductDao()
                instance = it }
            }
        }
    }

    override suspend fun insertProduct(product: Product): Long {
        return withContext(Dispatchers.IO) { dao.insert(product) }
    }
    override suspend fun deleteProduct(product: Product): Int {
        return withContext(Dispatchers.IO) { dao.delete(product) }
    }
    override suspend fun getAllStored(): List<Product> {
        return withContext(Dispatchers.IO) { dao.getAll() }
    }

}
